ddtrace/context.py

```diff
@@ -17,50 +20,40 @@
 log = get_logger(__name__)
 
 
+@attr.s(eq=True, slots=True)
 class Context(object):
-    """Represents the state required to propagate a trace across process
+    """Represents the state required to propagate a trace across execution
     boundaries.
     """
 
-    __slots__ = ["trace_id", "span_id", "_lock", "_span", "_meta", "_metrics"]
-
-    def __init__(
-        self,
-        trace_id=None,  # type: Optional[int]
-        span_id=None,  # type: Optional[int]
-        sampling_priority=None,  # type: Optional[NumericType]
-        dd_origin=None,  # type: Optional[str]
-    ):
-        # type: (...) -> None
-        self.trace_id = trace_id
-        self.span_id = span_id
-
-        self._lock = threading.RLock()
-        self._meta = {}  # type: _MetaDictType
-        self._metrics = {}  # type: _MetricDictType
-        self.dd_origin = dd_origin
-        self.sampling_priority = sampling_priority
-
-    def __eq__(self, other):
-        return (
-            self.span_id == other.span_id
-            and self.trace_id == other.trace_id
-            and self.sampling_priority == other.sampling_priority
-            and self.dd_origin == other.dd_origin
-            and self._meta == other._meta
-            and self._metrics == other._metrics
-        )
+    trace_id = attr.ib(default=None)  # type: Optional[int]
+    span_id = attr.ib(default=None)  # type: Optional[int]
+    _dd_origin = attr.ib(default=None)  # type: Optional[str]
+    _sampling_priority = attr.ib(default=None)  # type: Optional[NumericType]
+    _lock = attr.ib(factory=threading.RLock, eq=False)  # type: threading.RLock
+    _meta = attr.ib(factory=dict)  # type: _MetaDictType
+    _metrics = attr.ib(factory=dict)  # type: _MetricDictType
+
+    def __attrs_post_init__(self):
+        self.dd_origin = self._dd_origin
```

Comment:
almost feels wasteful to continue to store `self._dd_origin` if we just duplicate it into `self._meta` and then never touch it again.

Reply:
See this thread for context: https://github.com/DataDog/dd-trace-py/pull/1936#discussion_r626495583

Reply:
You could `del` it after if it pleases you :)

Reply:
Adding the `del` actually exposed an issue with the auto generated `__eq__` method: it uses `_dd_origin` and `_sampling_priority` which is wrong since it should be the ones in the `_meta` and `_metrics` dictionaries.